### PR TITLE
feat: add structureTranslationModule

### PR DESCRIPTION
**Pull request by RybakovaIE**

### CODE REVIEW

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+{
+
+SC_AGENT_IMPLEMENTATION(StructureTranslationAgent)
+{
+  ScAddr const & actionNode = otherAddr;
+  try
+  {
+    if (checkActionClass(actionNode) == SC_FALSE)
+      return SC_RESULT_OK;
+    SC_LOG_DEBUG("structureTranslationAgent started");
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    SC_LOG_INFO("structureTranslationAgent started");
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    if (m_memoryCtx.SetLinkContent(translationLink, translation) == SC_FALSE)
+      SC_THROW_EXCEPTION(utils::ScException, "structureTranslationAgent: cannot set link content");
+
+    ScAddrVector const & answerElements = {translationLink};
+
+    if (answerElements.empty())
+      SC_THROW_EXCEPTION(utils::ScException, "structureTranslationAgent: answer is empty");
+
+    utils::AgentUtils::finishAgentWork(&m_memoryCtx, actionNode, answerElements, true);
+    SC_LOG_DEBUG("structureTranslationAgent finished");
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    SC_LOG_INFO("structureTranslationAgent finished");
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    return SC_RESULT_OK;
+  }
+  catch (utils::ScException const & exception)
+  {
+    SC_LOG_ERROR(exception.Description());
+    utils::AgentUtils::finishAgentWork(&m_memoryCtx, actionNode, false);
+    SC_LOG_DEBUG("structureTranslationAgent finished with error");
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    SC_LOG_INFO("structureTranslationAgent finished with error");
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+SC_AGENT_IMPLEMENTATION(StructureTranslationAgent)
+{
+  ScAddr const & actionNode = otherAddr;
+  try
+  {
+    if (checkActionClass(actionNode) == SC_FALSE)
+      return SC_RESULT_OK;
+    SC_LOG_DEBUG("structureTranslationAgent started");
+
+    ScAddr const & nodeAddr = utils::IteratorUtils::getAnyByOutRelation(&m_memoryCtx, actionNode, scAgentsCommon::CoreKeynodes::rrel_1);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    ScAddr const & structuresSet = utils::IteratorUtils::getAnyByOutRelation(&m_memoryCtx, actionNode, scAgentsCommon::CoreKeynodes::rrel_1);
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    std::vector<std::string> translations;
+
+    auto const & structIterator = m_memoryCtx.Iterator3(nodeAddr, ScType::EdgeAccessConstPosPerm, ScType::NodeStruct);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    auto const & structIterator = m_memoryCtx.Iterator3(nodeAddr, ScType::EdgeAccessConstPosPerm, ScType::NodeConstStruct);
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    while (structIterator->Next())
+    {
+      ScAddr const & structAddr = structIterator->Get(2);
+      translations.push_back(translateStructure(structAddr, &m_memoryCtx));
+    }
+
+    std::string const & translation = join(translations.cbegin(), translations.cend(), TranslationConstants::STRUCTURES_DELIMITER);
+
+    ScAddr const & translationLink = m_memoryCtx.CreateLink();
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]: Add SC_LOG_DEBUG with translation content

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    if (translationLink.IsValid() == SC_FALSE)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    if (context.IsElement(translationLink) == SC_FALSE)
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      SC_THROW_EXCEPTION(utils::ScException, "structureTranslationAgent: cannot create answer link");
+    if (m_memoryCtx.SetLinkContent(translationLink, translation) == SC_FALSE)
+      SC_THROW_EXCEPTION(utils::ScException, "structureTranslationAgent: cannot set link content");
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
      SC_THROW_EXCEPTION(utils::ScException, "StructureTranslationAgent: cannot set link content");
```
Check everywhere

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+    ScAddrVector const & answerElements = {translationLink};
+
+    if (answerElements.empty())
+      SC_THROW_EXCEPTION(utils::ScException, "structureTranslationAgent: answer is empty");
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]: remove this if. It is never true

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    utils::AgentUtils::finishAgentWork(&m_memoryCtx, actionNode, false);
+    SC_LOG_DEBUG("structureTranslationAgent finished with error");
+    return SC_RESULT_ERROR;
+  }
+}
+
+bool StructureTranslationAgent::checkActionClass(ScAddr const & actionNode)
+{
+  return m_memoryCtx.HelperCheckEdge(
+      TranslationKeynodes::action_translate_structures, actionNode, ScType::EdgeAccessConstPosPerm);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]: action_translate_structures to what? Rename this action class to be clear, please

File: cxx/structureTranslationModule/keynodes/TranslationKeynodes.cpp [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#include "sc-memory/sc_memory.hpp"
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/keynodes/TranslationKeynodes.cpp:7]: add licence header about MIT to all c++ files

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+  ScAgentInit(true);
+  SC_AGENT_REGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+void deinitializeClasses()
+{
+  SC_AGENT_UNREGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+TEST_F(StructureTranslationTest, TestSetFillingWithPointers)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]: this test looks strange. It test c++ standard library?

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+TEST_F(StructureTranslationTest, TestNrelFromQuasybinaryNodeTranslator)
+{
+  ScMemoryContext & context = *m_ctx;
+  initializeClasses();
+
+  loader.loadScsFile(context, TEST_FILES_DIR_PATH + "testNrelFromQuasybinaryNodeTranslator.scs");
+  ScAddr test_structure = context.HelperFindBySystemIdtf(TEST_STRUCTURE_ALIAS);
+
+  auto translator = new NrelFromQuasybinaryNodeTranslator(&context);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]: I'd like to use simple object, without pointer, so you would not need to delete it. CHeck everywhere in tests

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+  EXPECT_EQ(answer.size(), 2u);
+  EXPECT_NE(std::find(answer.cbegin(), answer.cend(), "thing parts decomposition big part"), answer.cend());
+  EXPECT_NE(std::find(answer.cbegin(), answer.cend(), "thing parts decomposition little part"), answer.cend());
+
+  deinitializeClasses();
+  delete translator;
+}
+
+TEST_F(StructureTranslationTest, TestNrelFromNodeTranslator)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp:121]: You have so many duplicated code. Could you create a fubction with args as test scs filename and expected result. Or even create an expected result as sc-link in test scs file

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+using namespace structureTranslationModule;
+
+namespace StructureTranslationAgent
+{
+ScsLoader loader;
+std::string const & TEST_FILES_DIR_PATH = STRUCTURE_TRANSLATION_MODULE_TEST_SRC_PATH "/testStructures/StructureTranslation/";
+std::string const & TEST_QUESTION_NODE_ALIAS = "test_question_node";
+std::string const & TEST_STRUCTURE_ALIAS = "test_structure";
+int const WAIT_TIME = 3000;
+size_t const MAX_TRANSLATIONS = 30;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]: looks like unused

File: cxx/structureTranslationModule/translator/StructureTranslator.cpp [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+      ScType::EdgeAccessConstPosPerm,
+      scAgentsCommon::CoreKeynodes::nrel_main_idtf);
+}
+
+bool StructureTranslator::isEnglish(ScAddr const & node) const
+{
+  return context->HelperCheckEdge(TranslationKeynodes::lang_en, node, ScType::EdgeAccessConstPosPerm);
+}
+
+std::string StructureTranslator::getEnglishMainIdtf(ScAddr const & node) const
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.cpp]: There is method in CommonUtils getMainIdtf, use it and pass lang_en. Remove your getEnglishMainIdtf

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp [hunk 17]
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include "sc-memory/sc_memory.hpp"
+#include "sc-memory/sc_addr.hpp"
+
+namespace structureTranslationModule
+{
+using ScAddrSet = std::unordered_set<ScAddr, ScAddrHashFunc<uint32_t>>;
+
+class StructureTranslator
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.hpp:16]: It looks like translator utils, not translator itself

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  ScAddr const & actionNode = otherAddr;
+  try
+  {
+    if (checkActionClass(actionNode) == SC_FALSE)
+      return SC_RESULT_OK;
+    SC_LOG_DEBUG("structureTranslationAgent started");
+
+    ScAddr const & nodeAddr = utils::IteratorUtils::getAnyByOutRelation(&m_memoryCtx, actionNode, scAgentsCommon::CoreKeynodes::rrel_1);
+
+    std::vector<std::string> translations;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]: Use stringstream and operator `<<`, not vector and join. Ниже in the cycle будет `translations << translateStructure(structAddr, &m_memoryCtx)` вместо `translations.push_back(translateStructure(structAddr, &m_memoryCtx));`

File: cxx/structureTranslationModule/translator/FromConceptTranslator.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
+#include "FromConceptTranslator.hpp"
+
+namespace structureTranslationModule
+{
+FromConceptTranslator::FromConceptTranslator(ScMemoryContext * context)
+  : StructureTranslator(context)
+{
+}
+
+std::vector<std::string> FromConceptTranslator::translate(ScAddr const & structAddr) const
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/FromConceptTranslator.cpp]: Why do you need std::vector<std::string>? Could you replace it with just string or stringstream?

Comment (MksmOrlov) [reply cxx/structureTranslationModule/translator/FromConceptTranslator.cpp]: Change everywhere in module

File: cxx/structureTranslationModule/translator/FromConceptTranslator.cpp [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+  std::string const NODE_ALIAS = "_node";
+  std::string const CLASS_ALIAS = "_class";
+  std::string const EDGE_ALIAS = "_edge";
+  ScAddr classNode, node;
+
+  ScTemplate scTemplate;
+  scTemplate.Triple(ScType::NodeVarClass >> CLASS_ALIAS, ScType::EdgeAccessVarPosPerm >> EDGE_ALIAS, ScType::NodeVar >> NODE_ALIAS);
+  scTemplate.Triple(structAddr, ScType::EdgeAccessVarPosPerm, EDGE_ALIAS);
+  ScTemplateSearchResult searchResult;
+  context->HelperSearchTemplate(scTemplate, searchResult);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/FromConceptTranslator.cpp]: Checking only the edge is not correct (especially in other translators), you should check all the elements. To make it easy and effective you can use SmartTemplateSearch, just check all items with HelperCheckEdge from structure addr

File: cxx/structureTranslationModule/translator/NrelFromNodeTranslator.cpp [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+{
+}
+
+std::vector<std::string> NrelFromNodeTranslator::translate(ScAddr const & structAddr) const
+{
+  std::vector<std::string> translations;
+  std::string const NODE_ALIAS = "_node";
+  std::string const SOURCE_ALIAS = "_source";
+  std::string const NREL_ALIAS = "nrel";
+  std::string const EDGE_ALIAS = "_edge";
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/NrelFromNodeTranslator.cpp]: move this constants to the separate class, make it static and use in all the translators. It is a good memory optimisation

File: cxx/structureTranslationModule/translator/NrelFromQuasybinaryLinkTranslator.cpp [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+std::vector<std::string> NrelFromQuasybinaryLinkTranslator::translate(ScAddr const & structAddr) const
+{
+  std::vector<std::string> translations;
+  std::string const TUPLE_ALIAS = "_tuple";
+  std::string const NODE_ALIAS = "_node";
+  std::string const NREL_ALIAS = "_nrel";
+  std::string const EDGE_ALIAS = "_edge";
+  ScAddr tupleNode, node, nrelNode;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/NrelFromQuasybinaryLinkTranslator.cpp]: according to our codestyle, we should declare variables on separate line. We have clang code formatter, you can run it to automatically refactor all the code

File: cxx/structureTranslationModule/translator/NrelFromQuasybinaryLinkTranslator.hpp [hunk 23]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include "StructureTranslator.hpp"
+
+namespace structureTranslationModule
+{
+/**
+ * node _<= _nrel_relation:: {_[]};;
+ */
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/NrelFromQuasybinaryLinkTranslator.hpp]: It is so cool comments, add it to other translators!

File: cxx/structureTranslationModule/translator/StructureTranslatorSet.hpp [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+public:
+  explicit StructureTranslatorSet(ScMemoryContext * context);
+  
+  std::vector<std::string> translate(ScAddr structAddr)
+      const;
+
+  ~StructureTranslatorSet();
+
+private:
+  std::set<StructureTranslator *> const handlers;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslatorSet.hpp]:
```suggestion
  std::unordered_set<StructureTranslator *> const handlers;
```

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Optimisation changes are sooo important because we need fast replies in ask-ai chat

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+    std::string translation;
+
+    auto const & structIterator = m_memoryCtx.Iterator3(structuresSet, ScType::EdgeAccessConstPosPerm, ScType::NodeConstStruct);
+    while (structIterator->Next())
+    {
+      ScAddr const & structAddr = structIterator->Get(2);
+      translation += translateStructure(structAddr, &m_memoryCtx);
+    }
+
+    SC_LOG_DEBUG(translation);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    SC_LOG_DEBUG("StructureTranslationAgent: translation result is " << translation);
```

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+  {
+    if (checkActionClass(actionNode) == SC_FALSE)
+      return SC_RESULT_OK;
+    SC_LOG_INFO("StructureTranslationAgent started");
+
+    ScAddr const & structuresSet = utils::IteratorUtils::getAnyByOutRelation(&m_memoryCtx, actionNode, scAgentsCommon::CoreKeynodes::rrel_1);
+
+    std::string translation;
+
+    auto const & structIterator = m_memoryCtx.Iterator3(structuresSet, ScType::EdgeAccessConstPosPerm, ScType::NodeConstStruct);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]:
```suggestion
    ScIterator3Ptr const & structIterator = m_memoryCtx.Iterator3(structuresSet, ScType::EdgeAccessConstPosPerm, ScType::NodeConstStruct);
```
We use auto keyword only with difficult long types

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+    while (structIterator->Next())
+    {
+      ScAddr const & structAddr = structIterator->Get(2);
+      translation += translateStructure(structAddr, &m_memoryCtx);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.cpp]: += operator is not optimal for std::string. Use stringstream with << or string_view object

File: cxx/structureTranslationModule/agent/StructureTranslationAgent.hpp [hunk 28]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "sc-memory/kpm/sc_agent.hpp"
+
+/*
+ * This source file is part of an OSTIS project. For the latest info, see http://ostis.net
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/agent/StructureTranslationAgent.hpp]: includes and pragma statement should be after licence header comment, Somewhere you forgot to add licence (in cpp or hpp files). Check everywhere

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+  ScAgentInit(true);
+  SC_AGENT_REGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+void deinitializeClasses()
+{
+  SC_AGENT_UNREGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+void testTranslator(ScMemoryContext & context, auto & translator, std::string fileName, std::vector<std::string> answer_phrases)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]:
```suggestion
void testTranslator(ScMemoryContext & context, auto & translator, std::string fileName, std::vector<std::string> answerPhrases)
```
We don't use snake_case variable names, check in this file

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 30]
```diff
@@ -0,0 +1,208 @@
...
+{
+  initializeClasses();
+
+  loader.loadScsFile(context, TEST_FILES_DIR_PATH + fileName);
+  ScAddr test_structure = context.HelperFindBySystemIdtf(TEST_STRUCTURE_ALIAS);
+  auto const & answer = translator.translate(test_structure);
+
+  for (std::string phrase : answer_phrases)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]:
```suggestion
  for (std::string const & phrase : answer_phrases)
```
to not copy string object

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 31]
```diff
@@ -0,0 +1,208 @@
...
+  {
+    EXPECT_TRUE(answer.find(phrase) != std::string::npos) << phrase;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]: what is `<< phrase` for?

File: cxx/structureTranslationModule/translator/StructureTranslator.cpp [hunk 32]
```diff
@@ -0,0 +1,55 @@
...
+{
+  std::string content;
+  if (isEnglish(linkNode))
+    context->GetLinkContent(linkNode, content);
+  return content;
+}
+
+bool StructureTranslator::isInStructure(ScAddr const & structAddr, ScAddr const & elementAddr) const
+{
+  return elementAddr==structAddr or (context->GetElementType(elementAddr) == ScType::EdgeAccessConstPosPerm and context->GetEdgeSource(elementAddr)==structAddr) or context->HelperCheckEdge(structAddr, elementAddr, ScType::EdgeAccessConstPosPerm);
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.cpp]:
```suggestion
  return elementAddr == structAddr or (context->GetElementType(elementAddr) == ScType::EdgeAccessConstPosPerm and context->GetEdgeSource(elementAddr) == structAddr) or context->HelperCheckEdge(structAddr, elementAddr, ScType::EdgeAccessConstPosPerm);
```
You can use auto formatter

File: cxx/structureTranslationModule/translator/StructureTranslatorSet.cpp [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+    delete handler;
+}
+
+std::string StructureTranslatorSet::translate(ScAddr structAddr) const
+{
+  std::string answer;
+  for (auto const & handler : handlers)
+  {
+    std::string translations = handler->translate(structAddr);
+    answer += translations + "\n";
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslatorSet.cpp]: do not concatenate std strings

File: cxx/structureTranslationModule/translator/FromConceptTranslator.cpp [hunk 34]
```diff
@@ -0,0 +1,49 @@
...
+  {
+    classNode = searchResult[TranslationConstants::CLASS_ALIAS];
+    node = searchResult[TranslationConstants::NODE_ALIAS];
+    std::string const & classMainIdtf = utils::CommonUtils::getMainIdtf(context, classNode, {TranslationKeynodes::lang_en});
+    if (classMainIdtf.empty())
+      return ScTemplateSearchRequest::CONTINUE;
+    std::string const & nodeMainIdtf = utils::CommonUtils::getMainIdtf(context, node, {TranslationKeynodes::lang_en});
+    if (nodeMainIdtf.empty())
+      return ScTemplateSearchRequest::CONTINUE;
+    translations += nodeMainIdtf + " is " + classMainIdtf + ". ";
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/FromConceptTranslator.cpp]: do not concatenate strings, check everywhere

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+      const = 0;
+
+  size_t hashCode() const;
+
+  bool operator<(StructureTranslator const & other) const;
+
+protected:
+  ScMemoryContext * context;
+
+  std::string getEnglishContent(ScAddr const & linkNode) const;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.hpp]: seems unused or you can use CommonUtils

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp [hunk 36]
```diff
@@ -0,0 +1,53 @@
...
+
+  bool isInStructure(ScAddr const & structAddr, ScAddr const & elementAddr) const;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.hpp:34]: Use just context.HelperCheckEdge insted of it

File: cxx/structureTranslationModule/translator/NrelInQuasybinaryNodeTranslator.cpp [hunk 37]
```diff
@@ -0,0 +1,64 @@
...
+{
+NrelInQuasybinaryNodeTranslator::NrelInQuasybinaryNodeTranslator(ScMemoryContext * context)
+  : StructureTranslator(context)
+{
+}
+
+std::string NrelInQuasybinaryNodeTranslator::translate(ScAddr const & structAddr) const
+{
+  std::string translations;
+  ScAddr tupleNode, node, nrelNode;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/NrelInQuasybinaryNodeTranslator.cpp]: newline and type for each variable. Check everywhere

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp [hunk 38]
```diff
@@ -0,0 +1,53 @@
...
+
+
+private:
+  ScIterator5Ptr getNrelMainIdtfIterator(ScAddr const & node) const;
+
+  bool isEnglish(ScAddr const & node) const;
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.hpp]: seems unused or you can use CommonUtils

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp [hunk 39]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+struct StructureTranslatorCmp
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.hpp]:
```suggestion
struct StructureTranslatorComporator
```

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
And check unresolved please

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp (hunk 36 above)
Comment (RybakovaIE) [reply cxx/structureTranslationModule/translator/StructureTranslator.hpp:34]: this function can't be replaced by just context.HelperCheckEdge, several conditions must be checked here

File: cxx/structureTranslationModule/translator/StructureTranslator.cpp [hunk 40]
```diff
@@ -0,0 +1,42 @@
...
+{
+  return this->hashCode() < other.hashCode();
+}
+
+bool StructureTranslator::isInStructure(ScAddr const & structAddr, ScAddr const & elementAddr) const
+{
+  return elementAddr == structAddr or (context->GetElementType(elementAddr) == ScType::EdgeAccessConstPosPerm and context->GetEdgeSource(elementAddr) == structAddr) or context->HelperCheckEdge(structAddr, elementAddr, ScType::EdgeAccessConstPosPerm);
+}
+
+ScIterator5Ptr StructureTranslator::getNrelMainIdtfIterator(ScAddr const & node) const
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/translator/StructureTranslator.cpp]: Looks like unused. Remove it

File: cxx/structureTranslationModule/translator/StructureTranslator.hpp (hunk 17 above)
Comment (MksmOrlov) [reply cxx/structureTranslationModule/translator/StructureTranslator.hpp:16]: Rename it please if you agree

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
Rename it to structureTranslationModule. I Can't build it
image
Check previous conversations

File: cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp [hunk 41]
```diff
@@ -0,0 +1,211 @@
...
+  ScAgentInit(true);
+  SC_AGENT_REGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+void deinitializeClasses()
+{
+  SC_AGENT_UNREGISTER(structureTranslationModule::StructureTranslationAgent);
+}
+
+void testTranslator(ScMemoryContext & context, auto & translator, std::string fileName, std::vector<std::string> answerPhrases)
```
Comment (MksmOrlov) [inline cxx/structureTranslationModule/test/units/testStructureTranslationAgent.cpp]: do not use auto in parameters

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Review by MksmOrlov — APPROVED** [review]

**Pull request merged**